### ISSUE TITLE
Add API for reading the LOH size threshold

### ISSUE THREAD

**Issue by davidfowl** [dotnet/runtime#37693]
## Background and Motivation

There are several APIs that hardcode the 85K threshold to avoid landing data on the LOH (usually for ephemeral operations):

- https://github.com/dotnet/runtime/blob/e2ed3631b9a2a97de10991e4b548f63086cdc67a/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L31-L34
- 
https://github.com/dotnet/aspnetcore/blob/66dd4919dafcb21c57a923a661cdbb0ff351713d/src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorSourceDocument.cs#L16
- https://github.com/dotnet/aspnetcore/blob/76d197f0dd0d0835585d82540f35acde9db5a0d5/src/Middleware/ResponseCaching/src/Streams/StreamUtilities.cs#L13-L16

As it turns out, you can change this value https://docs.microsoft.com/en-us/dotnet/core/run-time-config/garbage-collector#large-object-heap-threshold. I'd love to update these locations to use the configured value (or less than that).

## Proposed API

cc @Maoni0

**Comment by Maoni0** [comment]
we should have an API that just reads config values in general - 

`int GetConfigValue (string configName);`

this may belong to some more general place than the GC class as we have other configs too. it can also make sense just being in the GC class as we know that GC configs are the most frequently used. 

I think I mentioned this to @stephentoub some time ago. any opinions?

**Comment by Maoni0** [comment]
> I know we had discussion about GC returning GC-specific configurations that would return all GC tuning arguments, both explicit ones and the discovered ones. I think that would be fine. I expect the primary use for this would be high-end perf diagnostic.

right, this is for folks who do high end perf diag and it's reasonable to expect them to know the names of the configs that they care about. so something like 

```csharp

struct ConfigInfo
{
    string name;
    // although we do have 2 GC configs with a string value - GCName and GCHeapAffinitizeRanges
    int value; 
};

static ConfigInfo[] GetConfigValues();
```

**Comment by davidfowl** [comment]
> For large object heap threshold specifically, I think it is anti-pattern to use large object heap threshold to take different paths or tune buffer sizes. It is ok to take use different implementations for small vs. big inputs, but the thresholds should not be based on GC tuning. It makes the GC tuning non-elastic and hard to reason about.

What should we do in practice? If we some of the allocations are ephemeral and end up on the LOH that could be problematic right? What would we do instead?

**Comment by jkotas** [comment]
It is unfortunate that there is a step function in the GC behavior around the LOH threshold. Let's assume that there would not be a step function like this and the change in behavior would be smooth as far as the GC is concerned (there are other performance step functions caused by cache sizes or thresholds in other subsystem).

How would you choose your buffer sizes for the best performance if there was no step function around the LOH threshold?

**Comment by davidfowl** [comment]
> How would you choose your buffer sizes for the best performance if there was no step function around the LOH threshold?

Test a variety of numbers sizes numbers based on 32/64 bit and measure the throughput (RPS) and working set and GC behavior, % time in GC, and number of collections. 

I think generally it comes down to a tradeoff between CPU and memory:
- Small buffers mean more operations and more CPU time
- Bigger buffers mean less operations, which results in potentially more memory waste and less operations

**Comment by jkotas** [comment]
Right, that's how we should be choosing the right thresholds. We should not assume that the LOH threshold is the right magic number.

> Small buffers mean more operations and more CPU time

Not necessarily. There are secondary effects:
- Small buffers have better cache location. Better cache locality means less CPU time.
- Small buffers are typically reused more. The cost of acquiring them is amortized better.

For example, let's say that somebody configured the LOH threshold to 32MB and the stream copying blindly used a chunk size that is close to LOH threshold. I would be easy to come up with examples that shows significant stream copying throughput regressions with the large LOH threshold.

**Comment by GSPP** [comment]
Maybe add `Bytes` so that it becomes `LargeObjectHeapThresholdBytes`? Other GC info APIs have been added recently where this was done.

> Having an explicit API for it, assuming we expect there's always going to be such a threshold, makes sense to me

In the future, the threshold could be different for different types. Or it could be dynamically adjusted at runtime. That would break the meaning of this exposed GC setting.

**Comment by davidfowl** [comment]
I still think we should add this API.

**Comment by GeeLaw** [comment]
#94068 (fixing #93963) exposes LOH threshold via `GetConfigurationVariables` as an `int`.

**Comment by janvorli** [comment]
Closing as it was implemented.
